### ISSUE TITLE
C# dynamic feature

### ISSUE THREAD

**Issue by eusebiu** [icsharpcode/ILSpy#154]
I've pasted only a part of the original and generated code but you can find it all in here: ICSharpCode.SharpDevelop.Project.WebProjectService.CreateVirtualDirectory

Original:
dynamic manager = webAdministrationAssembly.CreateInstance("Microsoft.Web.Administration.ServerManager");
if (manager.Sites[DEFAULT_WEB_SITE] != null) {
...
} else {...}

Generated:
object arg2 = assembly.CreateInstance("Microsoft.Web.Administration.ServerManager");
                            if (WebProjectService.<CreateVirtualDirectory>o__SiteContainer0.<>p__Site1 == null)
                            {
                                WebProjectService.<CreateVirtualDirectory>o__SiteContainer0.<>p__Site1 = CallSite<Func<CallSite, object, bool>>.Create(Microsoft.CSharp.RuntimeBinder.Binder.UnaryOperation(CSharpBinderFlags.None, ExpressionType.IsTrue, typeof(WebProjectService), new CSharpArgumentInfo[]
                                {
                                    CSharpArgumentInfo.Create(CSharpArgumentInfoFlags.None, null)
                                }));
                            }
                            Func<CallSite, object, bool> arg_1CB_0 = WebProjectService.<CreateVirtualDirectory>o__SiteContainer0.<>p__Site1.Target;
                            CallSite arg_1CB_1 = WebProjectService.<CreateVirtualDirectory>o__SiteContainer0.<>p__Site1;
                            if (WebProjectService.<CreateVirtualDirectory>o__SiteContainer0.<>p__Site2 == null)
                            {
                                WebProjectService.<CreateVirtualDirectory>o__SiteContainer0.<>p__Site2 = CallSite<Func<CallSite, object, object, object>>.Create(Microsoft.CSharp.RuntimeBinder.Binder.BinaryOperation(CSharpBinderFlags.None, ExpressionType.NotEqual, typeof(WebProjectService), new CSharpArgumentInfo[]
                                {
                                    CSharpArgumentInfo.Create(CSharpArgumentInfoFlags.None, null), 
                                    CSharpArgumentInfo.Create(CSharpArgumentInfoFlags.Constant, null)
                                }));
                            }

**Comment by dgrunwald** [comment]
Implementation idea: Replace call site delegate invocations with special IL instructions.

Create a bunch of new types of IL instructions, one for each method in the Microsoft.CSharp.RuntimeBinder.Binder class:
  * DynamicBinaryOperation
  * DynamicConvert
  * DynamicGetIndex
  * DynamicGetMember
  * DynamicInvoke
  * DynamicInvokeConstructor
  * DynamicInvokeMember
  * DynamicIsEvent
  * DynamicSetIndex
  * DynamicSetMember
  * DynamicUnaryOperation

These instructions would contain the information from the dynamic call-site, and have child-instructions for the arguments passed to the delegate call.
For example, `class DynamicBinaryOperation` would have fields:
```
CSharpBinderFlags flags;
ExpressionType operation;
IType context;
ImmutableArray<DynamicArgumentInfo> argumentInfo;
ILInstruction left;
ILInstruction right;
```

All the dynamic instruction types should probably derive from a common base class (`DynamicInstruction`?), and the `flags` member might be stored in that base class.

Create a new ILAst transform that detects the `if (cachedCallSite == null) { ... }` pattern and replaces `cachedCallSite.Target` delegate invocations with the new ILInstructions.
Due to interactions with short-circuiting operators `&&`/`||`, this transform should run before `ConditionDetection`. Maybe somewhere around the switch transforms?

The transform should probably not expect the call site initialization to be directly before the target delegate invocation -- while pattern held true in all our tests so far, I wouldn't count on it.

Extend the C# `ExpressionBuilder` to translate the new ILAst instructions into the corresponding C# expressions. It's important to cast all input operands to the exact expected type -- either the static type stored in the call-site, or to `dynamic` (also, mind special cases such as the `null` literal).

Now, some instructions (e.g. `DynamicIsEvent` or `DynamicUnaryOperation` with `operator false` don't have a direct C# equivalent). Instead, they are part of a higher-level pattern made up of multiple ILAst instructions (so far, we've identified "dynamic compound assignments" and "dynamic short-circuiting operators" as separate high-level patterns, but there might be more).
We'll handle these in a separate transform. But I think we can ignore them for the first implementation, as there are plenty uses of `dynamic` that won't encounter these special cases.

**Comment by siegfriedpammer** [comment]
Implemented in 58c174c87d4299068f5b54792a51421f4ffd56bf
